syntax = "proto3";

package nhdreport;

import "google/protobuf/timestamp.proto";

option go_package = "github.com/seans3/nhd/proto/gen/go;nhd_report";

// ========== User ==========
message Permissions {
  bool can_create_customers = 1;
  bool can_generate_reports = 2;
  bool is_admin = 3;
}
message User {
  string user_id = 1; // Firebase Auth UID
  string full_name = 2;
  string email = 3;
  Permissions permissions = 4;
  google.protobuf.Timestamp created_at = 5;
}

// ========== Customer ==========
message Customer {
  string customer_id = 1;
  string full_name = 2;
  string email = 3;
  string company_name = 4;
  google.protobuf.Timestamp created_at = 5;
  string created_by_user_id = 6;
}

// ========== Property Address ==========
message PropertyAddress {
  string property_address_id = 1;
  message AddressDetails {
    string street_address = 1;
    string street_address_2 = 2;
    string city = 3;
    string state = 4;
    string zip_code = 5;
    string zip_plus_4 = 6;
  }
  AddressDetails address_details = 2;
  message Coordinates {
    double latitude = 1;
    double longitude = 2;
  }
  Coordinates coordinates = 3;
  string plus_code = 4;
  string google_place_id = 5;
}

// ========== Report Run ==========
message ReportRun {
  string report_run_id = 1;
  string customer_id = 2;
  // The ID of the internal user who created the report run on behalf of a customer.
  // This field is optional. If it's not set, it implies the customer
  // (identified by customer_id) created the report for themselves.
  string created_by_user_id = 3;
  string property_address_id = 4;
  google.protobuf.Timestamp created_at = 5;
  enum Status {
    STATUS_UNSPECIFIED = 0;
    PENDING = 1;
    PROCESSING = 2;
    COMPLETED = 3;
    FAILED = 4;
  }
  Status status = 6;
  message HazardResults {
    bool in_special_flood_hazard_area = 1;
    bool in_dam_inundation_area = 2;
    bool in_very_high_fire_hazard_severity_zone = 3;
    bool in_wildland_fire_area = 4;
    bool in_earthquake_fault_zone = 5;
    bool in_seismic_hazard_zone = 6;
  }
  HazardResults results = 7;
  string template_reference = 8;
  string final_pdf_storage_path = 9;
  message EmailDelivery {
    enum DeliveryStatus {
      STATUS_UNSPECIFIED = 0;
      SENT = 1;
      FAILED = 2;
    }
    DeliveryStatus status = 1;
    google.protobuf.Timestamp sent_at = 2;
    string email_template_reference = 3;
  }
  repeated EmailDelivery email_deliveries = 10;
  bool disable_automatic_email = 11;

  // Financials
  message ReportCost {
    double amount = 1;
    string currency = 2; // e.g., "USD"
    google.protobuf.Timestamp set_at = 3;
    string set_by_user_id = 4;
  }
  repeated ReportCost cost_history = 12; // Complete, auditable history of cost changes.

  message Payment {
    enum PaymentStatus {
      PAYMENT_STATUS_UNSPECIFIED = 0;
      OUTSTANDING = 1;
      PAID = 2;
      REFUNDED = 3;
    }
    PaymentStatus status = 1;
    double amount_paid = 2;
    string currency = 3;
    google.protobuf.Timestamp paid_at = 4;
    string payment_method = 5; // e.g., "Stripe", "Manual"
    string transaction_id = 6;
  }
  Payment payment_details = 13;
}
